### PR TITLE
Add support for Nvidia Virtual GPU (vGPU)

### PR DESCRIPTION
**Pull request by lnguyen**
# Description

Nvidia [Virtual GPU](https://docs.nvidia.com/grid/latest/grid-vgpu-release-notes-vmware-vsphere/index.html) (vGPU), formerly Nvidia Grid, allows hardware GPU resources to be divided among several VMs. For example, a 16 GB Nvidia T4 can be divided among 4 VMs with 4 GB GPU memory apiece. GPUs are typically used for AI-related applications such as LLMs (Large Language Models).

This commit enables Nvidia vGPU support for the BOSH vSphere CPI. vGPUs are configured via the `cloud_properties` property, `vgpus`, which is an array of vGPUs. When `vgpus` are configured, the properties, `memory_reservation_locked_to_max` and `upgrade_hw_version` are additionally to `true`. The former is a requirement for attaching a vGPU to a VM, the latter, for multiple vGPUs attached to a VM.

A typical vGPU device name is `grid_t4-16q`, which is an Nvidia T4 card with 16 GB of RAM. The devices not ending in `q` don't seem to work (e.g. `grid_t4-1a` doesn't work).

Here's a typical VM Extension:

```yaml
vm_extensions:
- name: vgpu
  cloud_properties:
    vgpus:
    - grid_t4-1q
```

There are unit tests, but we didn't write integration tests because we don't have access to VMware hardware with GPUs.

vGPUs requires the installation of special ESXi VIBs (vSphere Installation Bundle).

## Type of change

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- We wrote unit tests
- We tested deploying VMs with vGPUs on a vSphere 8.0 environment with an Nvidia T4 PCI card.
- We did **not** write integration tests; we don't have access to a VMware environments with GPUs.

### CODE REVIEW

**Review by cunnie — COMMENTED** [review]
I love this PR!

- ~~I prefer the old property name, `pci_passthrough_device_id`, over the newer one, `vgpus`, because "PCI Passthrough" is what the users see on the web interface, but I suspect you had pretty good reasons to change it. Let's talk on Monday.~~ I agree, `vgpus` is a better name.
- I like that it's an array ~~even though you can only have one~~.
- I'd love an integration test, but I don't think that's possible. 

BOSH.IO docs

- **vgpus** [Array, optional]: Specifies an Nvidia GRID vGPU to attach to VM.  Automatically sets the properties  `memory_reservation_locked_to_max` and `upgrade_hw_version` to `true`. The former property is a requirement, the latter is necessary for > 1 vGPU. Available in v97+

```yaml
vm_extensions:
- name: gpu
  vgpus:
   - grid_t4-16q
```

Notes to self:
- If I deploy 8 VMs with `vgpus [ 'grid_t4-1q' ]`, does vSphere place all 8 VMs on the host that has the PCI passthrough? **Yes**. The 8 VMs deploy successfully, each with a PCI Device ` grid_t4-1q `.
- If another VM has consumed the sole "grid_t4-16q", what happens when we deploy a 2nd VM?
```
Task 218618 | 12:13:21 | Creating missing vms: vgpu/401dbe13-26cd-4b49-a846-c1febce4c070 (0) (00:00:12)
                       L Error: Unknown CPI error 'Unknown' with message 'Could not power on VM '<[Vim.VirtualMachine] vm-29699>': No host is compatible with the virtual machine.' in 'create_vm' CPI method (CPI request ID: 'cpi-620112')
```
- If I deploy a VM with a non-existent vgpu `vgpus [ 'non-existent' ]`, what happens?
```
Task 218631 | 13:42:54 | Creating missing vms: vgpu/eec114e6-de32-461c-a724-9ce38f5d65e9 (0) (00:00:12)
                       L Error: Unknown CPI error 'Unknown' with message 'Could not power on VM '<[Vim.VirtualMachine] vm-29736>': No host is compatible with the virtual machine.' in 'create_vm' CPI method (CPI request ID: 'cpi-708847')
```
- If I deploy a VM with eight vgpus `vgpus [ 'grid_t4-1q', ..., 'grid_t4-1q' ]`, what happens? Success.
- Is `upgrade_hw_version: true` necessary for multiple vGPUs on one VM? **Yes**. When it's missing we get the following error during `bosh deploy`:
```
Task 218632 | 13:44:53 | Creating missing vms: vgpu/eec114e6-de32-461c-a724-9ce38f5d65e9 (0) (00:01:54)
                       L Error: Unknown CPI error 'Unknown' with message ''2' of device 'vmiop' was found. Only '1' allowed' in 'create_vm' CPI method (CPI request ID: 'cpi-763637')
```
- Is `memory_reservation_locked_to_max: true` necessary? **Yes**. When it's missing we get the following error during `bosh deploy`:
```
Task 218624 | 13:18:55 | Creating missing vms: vgpu/687ea88d-acbc-45e4-9218-3f340f8b7bac (1) (00:00:31)
                       L Error: Unknown CPI error 'Unknown' with message 'Invalid memory setting: memory reservation (sched.mem.min) should be equal to memsize(8192). ' in 'create_vm' CPI method (CPI request ID: 'cpi-610211')
```

**Review by cunnie — APPROVED** [review]
Great work, @lnguyen !

**Comment by jpalermo** [comment]
Do we think vsphere may allow multiple vgpus in the future? If not, an array that can't have more than one value seems odd.

Also, a required property of `memory_reservation_locked_to_max` seems strange. Shouldn't we just set that automatically on the CPI side of things if you are asking for a vgpu?

`cloud_properties` are CPI specific items, but they don't have to be dictated by the IaaS interface, the CPI should do processing on the `cloud_properties` if it makes a cleaner interface for the operator.

**Comment by cunnie** [comment]
> Do we think vsphere may allow multiple vgpus in the future? If not, an array that can't have more than one value seems odd.

Surprisingly, vSphere **allows multiple vgpus** (it works when I set them manually from the GUI), but vSphere throws an error when we try to set multiple GPUs via the CPU. I haven't figured out why, but I don't know how important it is to fix this CPI bug—I can't envision a scenario where a user would want multiple GPUs.

> Also, a required property of memory_reservation_locked_to_max seems strange. Shouldn't we just set that automatically on the CPI side of things if you are asking for a vgpu?

Probably a good idea, and not to hard to implement—let me go back to the drawing board.

**Review by jpalermo — APPROVED** [review]
👍

File: src/vsphere_cpi/lib/cloud/vsphere/vm_creator.rb [hunk 1]
```diff
@@ -110,6 +110,13 @@ def create(vm_config)
               config_spec.device_change << nic_config
             end
           end
+          unless vm_config.vgpus.empty?
+            vm_config.vgpus.each do |vgpu|
```
[the remark this answers is not in the document]
Comment (cunnie) [reply src/vsphere_cpi/lib/cloud/vsphere/vm_creator.rb]: Done!

**Comment by wyike** [comment]
> > Do we think vsphere may allow multiple vgpus in the future? If not, an array that can't have more than one value seems odd.
> 
> Surprisingly, vSphere **allows multiple vgpus** (it works when I set them manually from the GUI), but vSphere throws an error when we try to set multiple GPUs via the ~CPU~ CPI. I haven't figured out why, but I don't know how important it is to fix this CPI bug—I can't envision a scenario where a user would want multiple GPUs.

In some doc like [this](https://www.yellow-bricks.com/2020/02/01/adding-multiple-vgpus-to-a-virtual-machine/), I see requirements/constrains to use multiple vGPU for a VM

I thought user may have scenarios to deploy multiple workloads and assign different GPUs to each on a single cluster/worker, is that not true?

**Comment by cunnie** [comment]
@wyike

> I thought user may have scenarios to deploy multiple workloads and assign different GPUs to each on a single cluster/worker, is that not true?

That's a good point — user might have scenarios to deploy multiple workloads on a single worker and assign different GPUs. And thanks for finding that [doc](https://www.yellow-bricks.com/2020/02/01/adding-multiple-vgpus-to-a-virtual-machine/). 

If you have users who need this feature, let us know and we can see if it we can revisit this.

**Pull request merged**